src/app/archive/lib/processor.ml

```diff
@@ -3547,12 +3547,49 @@ let add_block_aux ?(retries = 3) ~logger ~pool ~add_block ~hash
     [%log info]
       "Populating token owners table for block with state hash $state_hash"
       ~metadata:[ ("state_hash", Mina_base.State_hash.to_yojson state_hash) ] ;
-    List.iter tokens_used ~f:(fun (token_id, owner) ->
-        match owner with
-        | None ->
-            ()
-        | Some acct_id ->
-            Token_owners.add_if_doesn't_exist token_id acct_id ) ;
+    let module Node = struct
+      type t = Token_id.t * Account_id.t option
+      [@@deriving equal, compare, hash, sexp]
+    end in
+    (* there's an edge from token A to token B if the B's token owner, an account identifier,
+       has A as its token; token A is added before token B
+    *)
+    let token_edges =
+      List.filter_map tokens_used
+        ~f:(fun ((token_id, acct_id_opt) as token_and_owner) ->
+          match acct_id_opt with
+          | None ->
+              assert (Token_id.equal token_id Token_id.default) ;
+              (* no edge to default token *)
+              None
+          | Some acct_id -> (
+              let token_id' = Account_id.token_id acct_id in
+              match
+                List.find tokens_used ~f:(fun (tokid, _) ->
```

Comment:
Nit: might be more efficient to build a map once at the beginning and lookup over that (`O(n log n)` vs `O(n^2)`)